cranelift/codegen/src/isa/aarch64/lower.isle

```diff
@@ -695,3 +695,222 @@
       (value_reg (alu_rs_imm_logic (ALUOp.EorNot64) $I64 x y)))
 
 (rule (lower (has_type $I128 (bxor_not x y))) (i128_alu_bitop (ALUOp.EorNot64) x y))
+
+;;;; Rules for `ishl` ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+;; Shift for i8/i16/i32.
+(rule (lower (has_type (fits_in_32 ty) (ishl x y)))
+      (value_reg (do_shift (ALUOp.Lsl32) ty (put_in_reg x) y)))
+
+;; Shift for i64.
+(rule (lower (has_type $I64 (ishl x y)))
+      (value_reg (do_shift (ALUOp.Lsl64) $I64 (put_in_reg x) y)))
+
+;; Shift for i128.
+(rule (lower (has_type $I128 (ishl x y)))
+      (lower_shl128 (put_in_regs x) (value_regs_get (put_in_regs y) 0)))
+
+;;     lsl     tmp3, src_lo, amt
+;;     lsl     tmp2, src_hi, amt
+;;     mvn     inv_amt, amt
+;;     lsr     tmp1, src_lo, #1
+;;     lsr     tmp1, tmp1, inv_amt
+;;     tst     amt, #0x40
+;;     orr     tmp2, tmp2, tmp1
+;;     csel    dst_hi, tmp3, tmp2, ne
+;;     csel    dst_lo, xzr, tmp3, ne
+(decl lower_shl128 (ValueRegs Reg) ValueRegs)
+(rule (lower_shl128 src amt)
+      (let (
+          (src_lo Reg (value_regs_get src 0))
+          (src_hi Reg (value_regs_get src 1))
+          (lo_lshift Reg (alu_rrr (ALUOp.Lsl64) src_lo amt))
+          (hi_lshift Reg (alu_rrr (ALUOp.Lsl64) src_hi amt))
+          (inv_amt Reg (alu_rrr (ALUOp.OrrNot32) (zero_reg) amt))
+          (lo_rshift Reg (alu_rrr (ALUOp.Lsr64)
+                                  (alu_rr_imm_shift (ALUOp.Lsr64)
+                                                    src_lo
+                                                    (imm_shift_from_u8 1))
+                                  inv_amt))
+
+          ;; TODO: this is probably best-expressed with something along the
+          ;; lines of `ProducesFlags` and `ConsumesFlags` but for now this
+          ;; should work at least while we transition.
```

Comment:
I think it'd probably be good to build out the `ProducesFlags` / `ConsumesFlags` type wrappers initially, just to avoid the cost later of undoing the "temporary manually-managed-for-now flags usage" approach. It's hopefully not more than 10-20 lines of ISLE? If it turns out to be too difficult for some reason then of course we can reconsider...

Reply:
Sure thing! I believe the most recent commit should take care of this